File: examples/ex1.ts
import init from '../src/webgpu-tensors'

const t = await init();
let x = t.rand([5, 3]);
await t.print(x);

x = t.ones([5, 3]);
await t.print(x);

x = t.rand([5, 3]);
await t.print(x);

x = t.tensor([0, 1, 2, 3]);
await t.print(x);

x = t.tensor([[0, 1, 2, 3],[3, 4, 5, 6]]);
await t.print(x);

x = t.tensor([[[0, 1], [2, 3]],[[3, 4], [5, 6]]]);
await t.print(x);

x = t.tensor([[[[0], [1]], [[2], [3]]],[[[3], [4]], [[5], [6]]]]);
await t.print(x);

const tensor = t.rand([3,4]);
t.print(`Shape of tensor: ${tensor.shape}`);
t.print(`Datatype of tensor: ${tensor.dtype}`);
t.print(`Device tensor is stored on: ${tensor.device}`);
await t.print("tensor", tensor, x);

x = t.empty([3, 4, 5]);
t.print(x.size());
t.print(x.size(1));

t.reset();
x = t.tensor([[0, 0, 0, 0], [1, 1, 1, 1], [2, 2, 2, 2], [3, 3, 3, 3]]);
const y = t.tensor([[0, 0, 0, 0], [1, 1, 1, 1], [2, 2, 2, 2], [3, 3, 3, 3]]);
await t.print('matmul', x, y);
const result = t.matmul(x, y);
t.print('matmul', result);

x = t.tensor([[-3, -2, -1, 0], [0, 1, 2, 3], [0, 1, 2, 3], [0, 1, 2, 3]]);
const max = t.maximum(x, 0);
await t.print("maximum=", max);

// Test for maximum function
const testTensor = t.tensor([[-3, -2, -1, 0], [0, 1, 2, 3], [0, 1, 2, 3], [0, 1, 2, 3]]);
const mul = t.matmul(testTensor, x);
const maxResult = t.maximum(mul, 0);
await t.print("Test maximum=", maxResult);

File: examples/ex1.rs
use webgpu_tensors::{RSTensors, Tensors, Tensor, tensors_println, Device};

fn main() {
    let mut t = RSTensors::new(Device::CPU);
    t.init(None).unwrap();

    // Rest of the code remains the same
    let x = t.rand(vec![5, 3], None);
    tensors_println!("{}", x);

    let x = t.ones(vec![5, 3], None);
    tensors_println!("{}", x);

    let x = t.rand(vec![5, 3], None);
    tensors_println!("{}", x);

    let x = t.tensor(vec![0.0, 1.0, 2.0, 3.0], Some(vec![4]), None);
    tensors_println!("{}", x);

    let x = t.tensor(vec![0.0, 1.0, 2.0, 3.0, 3.0, 4.0, 5.0, 6.0], Some(vec![2, 4]), None);
    println!("{}", x);

    let x = t.tensor(vec![0.0, 1.0, 2.0, 3.0, 3.0, 4.0, 5.0, 6.0], Some(vec![2, 2, 2]), None);
    println!("{}", x);

    let x = t.tensor(vec![0.0, 1.0, 2.0, 3.0, 3.0, 4.0, 5.0, 6.0], Some(vec![2, 2, 2, 1]), None);
    println!("4D tensor:");
    println!("{}", x);

    let tensor = t.rand(vec![3, 4], None);
    println!("Shape of tensor: {:?}", tensor.shape());
    println!("Datatype of tensor: {:?}", tensor.dtype());
    println!("Device tensor is stored on: {:?}", tensor.device());
    println!("tensor: {}", tensor);

    let x = t.empty(vec![3, 4, 5], None);
    println!("Size of empty tensor: {}", x.numel());
    println!("Size of dimension 1: {}", x.size(Some(1)).unwrap());

    t.reset();
    let x = t.tensor(vec![0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0], Some(vec![4, 4]), None);
    let y = t.tensor(vec![0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0], Some(vec![4, 4]), None);
    println!("Matrix multiplication:");
    println!("x: {}", x);
    println!("y: {}", y);
    let result = t.matmul(&x, &y);
    println!("Result: {}", result);

    let x = t.tensor(vec![-3.0, -2.0, -1.0, 0.0, 0.0, 1.0, 2.0, 3.0, 0.0, 1.0, 2.0, 3.0, 0.0, 1.0, 2.0, 3.0], Some(vec![4, 4]), None);
    let max = t.maximum(&x, 0.0);
    println!("Maximum (element-wise) with 0:");
    println!("{}", max);

    // Test for maximum function
    let test_tensor = t.tensor(vec![-3.0, -2.0, -1.0, 0.0, 0.0, 1.0, 2.0, 3.0, 0.0, 1.0, 2.0, 3.0, 0.0, 1.0, 2.0, 3.0], Some(vec![4, 4]), None);
    let mul = t.matmul(&test_tensor, &x);
    let max_result = t.maximum(&mul, 0.0);
    println!("Test maximum:");
    println!("{}", max_result);

    t.destroy();

    // Now test with GPU
    let mut t = RSTensors::new(Device::GPU);
    t.init(None).unwrap();

    let x = t.rand(vec![5, 3], None);
    tensors_println!("GPU tensor: {}", x);

    t.destroy();
}
